include/swift/SIL/SILDifferentiabilityWitness.h

```diff
@@ -0,0 +1,156 @@
+  /// devirtualization from another module.
+  bool serialized;
+  /// The AST `@differentiable` or `@differentiating` attribute from which the
+  /// differentiability witness is generated. Used for diagnostics.
+  /// Null if the differentiability witness is parsed from SIL or if it is
+  /// deserialized.
+  DeclAttribute *attribute = nullptr;
+
+  static AutoDiffConfig *
+  getAutoDiffConfig(SILModule &module, IndexSubset *parameterIndices,
+                    IndexSubset *resultIndices,
+                    GenericSignature *derivativeGenSig);
+
+  SILDifferentiabilityWitness(SILModule &module, SILLinkage linkage,
+                              SILFunction *originalFunction,
+                              IndexSubset *parameterIndices,
+                              IndexSubset *resultIndices,
+                              GenericSignature *derivativeGenSig,
+                              SILFunction *jvp, SILFunction *vjp,
+                              bool isSerialized, DeclAttribute *attribute)
+    : module(module), linkage(linkage), originalFunction(originalFunction),
+      parameterIndices(parameterIndices), resultIndices(resultIndices),
+      derivativeGenericSignature(derivativeGenSig), jvp(jvp), vjp(vjp),
+      serialized(isSerialized), attribute(attribute) {}
+
+public:
+  static SILDifferentiabilityWitness *create(
+      SILModule &module, SILLinkage linkage, SILFunction *originalFunction,
+      IndexSubset *parameterIndices, IndexSubset *resultIndices,
+      GenericSignature *derivativeGenSig, SILFunction *jvp, SILFunction *vjp,
+      bool isSerialized, DeclAttribute *attribute = nullptr);
+
+  SILDifferentiabilityWitnessKey getKey() const;
+  SILModule &getModule() const { return module; }
+  SILLinkage getLinkage() const { return linkage; }
+  SILFunction *getOriginalFunction() const { return originalFunction; }
+  IndexSubset *getParameterIndices() const {
+    return parameterIndices;
+  }
+  IndexSubset *getResultIndices() const {
+    return resultIndices;
+  }
+  GenericSignature *getDerivativeGenericSignature() const {
+    return derivativeGenericSignature;
+  }
+  SILFunction *getJVP() const { return jvp; }
+  SILFunction *getVJP() const { return vjp; }
+  SILFunction *getDerivative(AutoDiffDerivativeFunctionKind kind) const {
+    switch (kind) {
+    case AutoDiffDerivativeFunctionKind::JVP: return jvp;
+    case AutoDiffDerivativeFunctionKind::VJP: return vjp;
+    }
+  }
+  void setDerivative(AutoDiffDerivativeFunctionKind kind,
+                     SILFunction *derivative) {
+    switch (kind) {
+    case AutoDiffDerivativeFunctionKind::JVP: jvp = derivative; break;
+    case AutoDiffDerivativeFunctionKind::VJP: vjp = derivative; break;
+    }
+  }
+  bool isSerialized() const { return serialized; }
+  DeclAttribute *getAttribute() const { return attribute; }
+
+  /// Verify that the differentiability witness is well-formed.
+  void verify(const SILModule &M) const;
```

Comment:
Stick to one naming convention. I see that you are using mostly camel case in this file, so it's better to make that consistent. Same for `print`'s arguments below.

```suggestion
  void verify(const SILModule &module) const;
```

Reply:
Fair. I will choose not to update the variable names in the defining files (`SILVerifier.cpp`, `SILPrinter.cpp`) to match the local formatting of those files.